[Huang Zhizhou] iP

huangzz125 opened the pull request:
# Grand Duke
> “The bad news is time flies. The good news is you’re the pilot.”

## Features
- Add tasks
- List tasks
- Mark and unmark tasks
- Delete tasks
- Search for tasks
- Tag tasks

## Usage
1. Download the app from [Releases](https://github.com/huangzz125/ip/releases)
2. Run the app
3. Add a task

### Example  - _it's that simple!_ 
```
todo go to the gym
```

## Main Method

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

## Upcoming Features
- [x] Simple Task Management
- [x] GUI
- [x] More Exceptions! :sweat_smile:

Commit:
Add Gradle support

huangzz125 committed:
Added greet, echo, exit features (Level 1)

huangzz125 committed:
Added list feature (Level 2)

huangzz125 committed:
Added the ability to mark tasks as done and undone (Level 3)

huangzz125 committed:
Added support for tracking of todo, deadline, event tasks

huangzz125 committed:
Added UI Testing (A-TextUiTesting)

huangzz125 committed:
Handled exceptions when users enter incorrect inputs (Level 5)

huangzz125 committed:
Added delete functionality (Level 6)

huangzz125 committed:
Add storage feature

huangzz125 committed:
Change strings of dates into LocalDate objects

huangzz125 committed:
Resolve merge conflicts with master

huangzz125 committed:
Fix minor bug

huangzz125 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

huangzz125 committed:
Add TextUi class and methods

huangzz125 committed:
Add TaskList class and methods

huangzz125 committed:
Add Storage class and its methods

huangzz125 committed:
Refine code in Storage class

huangzz125 committed:
Add parser class and its methods

huangzz125 committed:
Add all the classes into duke package

huangzz125 committed:
Add 2 tests for Commands

huangzz125 committed:
Modify Storage file name for Jar to work

huangzz125 committed:
Format code to adhere to coding standard

huangzz125 committed:
Add JavaDocs

huangzz125 committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

huangzz125 committed:
Shorten lines that exceeds character limit

huangzz125 committed:
Add find feature

huangzz125 committed:
Add Checkstyle files

huangzz125 committed:
Merge branch 'branch-A-CodingStandard'

huangzz125 committed:
Add minor checkstyle edits

huangzz125 committed:
Merge branch 'master' into branch-A-CodingStandard

huangzz125 committed:
Touch up on checkstyle

huangzz125 committed:
Implement a simple GUI

TamKengHong commented on data.txt:
```diff
@@ -0,0 +1 @@
+E,0,project meeting,2000-12-20,2000-12-20
```
What is the difference between data.txt and duke.txt?

huangzz125 replied:
I forgot to delete duke.txt and its folder. Good spot and thanks!

TamKengHong commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+/**
+ * This class contains the Commands for Duke.
+ */
+public class Commands {
```
Good and well organised!

TamKengHong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,27 @@
+        TextUi.getWelcomeMessage();
+
+        String input;
+        while (!(input = TextUi.in.nextLine()).equals("bye")) {
```
Cool way of taking in input!

TamKengHong commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public String toStorableString() {
+        return "E" + "," + (this.isDone() ? "1" : "0") + "," + this.getDescription()
```
Suggestion: Use String.format() instead of doing "" + "" + "" + "" + "", looks clearer

TamKengHong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,64 @@
+     * @param textUi    TextUi.
+     */
+    public void parse(
+            String userInput, TaskList taskList, Storage storage, TextUi textUi)
```
Indentation: this line should be together with the line on top

TamKengHong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+     * @return Task.
+     */
+    public Task parseStringToTask(String string) {
+        String[] arr = string.split(",");
```
Suggestion: string and arr are ambiguous, can use String input and inputArr instead.

TamKengHong reviewed (commented):
Overall well done! Good code and variable naming, methods were clear and code was well organised.

huangzz125 committed:
Take up suggestions made in PR review

huangzz125 committed:
Integrate my Duke into the GUI

eugenetangkj commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+     * @param taskList TaskList for Duke.
+     * @return String
+     */
+    public static String executeListCommand(TaskList taskList) {
```
Good naming of variable `taskList`.

eugenetangkj commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+     * @param textUi   TextUi for Duke.
+     * @param taskList TaskList for Duke.
+     */
+    public static String executeFindCommand(String toFind, TextUi textUi, TaskList taskList) {
```
Would `toFind` be unreflective of what is actually to be found? A suggestion would be `commandToFind` or `searchTerm` which is more explicit in its meaning.

eugenetangkj commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+
+    /**
+     * Executes the mark command.
+     * @param input    input String.
```
I notice the extra spacing between the variable name `input` and its description `input String`. I believe that your intention is to make it align with the subsequent Javadoc comments. Good effort as it makes it neater, but it might not be required according to the example shown [here](https://nus-cs2103-ay2223s2.github.io/website/book/documentation/tools/javaDoc/how/).

huangzz125 replied:
You are right, thanks.

eugenetangkj commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    /**
+     * Executes the todo command.
```
Should you add a line between the description and the subsequent parts of the Javadoc comments?

eugenetangkj commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,186 @@
+     */
+    public static String executeDeadlineCommand(String input, TextUi textUi, TaskList taskList, Storage storage)
+            throws DukeException {
+        String check = Parser.removeWhiteSpaces(input);
```
Should you be naming variables using nouns instead of verbs? For example, `check` could be better named as `processedString`.

eugenetangkj commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDate by;
```
The variable name `by` might not be very clear to readers regarding what value it holds. A suggestion would be `deadline`.

eugenetangkj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,137 @@
+package duke;
+
+import javafx.application.Application;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
```
Great job in importing only classes that you need, instead of using a wildcard import. 👍

eugenetangkj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,54 @@
+     * @param storage   Storage.
+     * @param textUi    TextUi.
+     */
+    public String parse(String userInput, TaskList taskList, Storage storage, TextUi textUi)
```
Good naming of the parameters names. I can easily understand what are they.

eugenetangkj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,54 @@
+            body = "";
+        }
+
+        switch (command) {
```
Nice! You have inserted correct indentation for the `switch` and `case` statements, avoiding a commonly-made mistake.

eugenetangkj commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,53 @@
+public abstract class Task {
+
+    private String description;
+    private boolean isDone;
```
Good naming of boolean variable `isDone`. 👍

eugenetangkj commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,53 @@
+        this.isDone = isDone;
+    }
+
+    public abstract String toStorableString();
```
Would a better name be `convertToStorableString`?

huangzz125 replied:
Nice suggestion!

eugenetangkj commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,53 @@
+        return this.isDone;
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    public void markAsUndone() {
+        this.isDone = false;
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
```
Good usage of verbs for naming of methods.

eugenetangkj reviewed (commented):
Great job in adhering to coding standards and code quality guidelines. Code is user-friendly and easy to understand. Just some nits to fix. Overall, neat and tidy implementation. 😀

huangzz125 committed:
Fix code to adhere to code standard after PR review 2

huangzz125 committed:
Modify application className

huangzz125 committed:
Add some Java assertions

There are portions of code in Commands, Parser & Storage classes that can make use of assertions.

By using these assert statements, we can define assumptions about the program state at certain junctures so that the runtime can verify them.

Let's identify these areas of correct where assertions can be beneficial to Grand Duke.

huangzz125 committed:
Merge pull request #1 from huangzz125/A-Assertions

Add some Java assertions

huangzz125 committed:
Enforce code quality

There are some portions of code in all the classes that may not be of the best code quality.
By critically examining through the code and enforce code quality, source code of Grand Duke can be more readable.
Let's identify these problems and fix them early.

huangzz125 committed:
Merge pull request #2 from huangzz125/A-CodeQuality

Enforce code quality

huangzz125 committed:
Add tag task feature

huangzz125 committed:
Merge pull request #3 from huangzz125/branch-Add-Tags

Add tag task feature

huangzz125 committed:
Edit JavaDoc for methods to adhere to standard

huangzz125 committed:
Implement an initial version of FXML

huangzz125 committed:
Enhance the Gui

huangzz125 committed:
Merge pull request #5 from huangzz125/A-BetterGui

A better Gui

huangzz125 committed:
Change some output strings in textUi to give Duke a personality

huangzz125 committed:
Add a few more exceptions

huangzz125 committed:
Remove Duke Exception

huangzz125 committed:
Merge pull request #6 from huangzz125/A-MoreErrorHandling

More Error Handling

huangzz125 committed:
Delete away an unused Exception

huangzz125 committed:
Finalise the GUI

huangzz125 committed:
Edit interface

huangzz125 committed:
Merge pull request #7 from huangzz125/A-Release

Edit GUI for final release

huangzz125 committed:
Fix delete bug

huangzz125 committed:
Make final changes before release

huangzz125 committed:
Fix bug on displayPicture

huangzz125 committed:
Fix readme.md typo

huangzz125 committed:
Add description to user guide
